### ISSUE TITLE
Can your pipeline be used for paired end reads with gaps?

### ISSUE THREAD

**Issue by guodudou2** [crukci-bioinformatics/ampliconseq#1]
Hello,

I am interested to confirm an allele frequency for a SNV in BRAF gene using amplicon sequencing data. The length of amplicon is 271bp, and paired reads are 101X2. So there is a gap between them. I am wondering whether your pipeline can be used in my case? I see your input is one bam file, what tools/steps do you recommend to align the fastq files to reference genome?

Thank you very much,
Wendy

**Comment by eldrid01** [comment]
Hi Wendy,

Yes, I believe it can although I don't think any of the panels we use contains an amplicon where the reads do not overlap. Clearly you won't be able to detect SNVs in the gap in the middle of the amplicon between the reads. Using this pipeline may be overkill for what you need, particularly if you're just looking at a single amplicon in just one sample. You could in the first instance inspect your aligned sequences in IGV. Your depth will probably be quite high in the region of interest so it may help to turn off down-sampling of reads in IGV (View > Preferences menu item, alignment tab).

We use bwa-mem2 for alignment and for variant calling generally choose a version of the reference genome that includes all the unplaced and unlocalized contigs as well as the decoy sequence, e.g. https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids, to try to avoid misplaced/misaligned reads causing false positive calls. For amplicon sequencing, this is probably less of an issue.

Hope that helps,
Matt

**Comment by guodudou2** [comment]
Hi Matt,

I really appreciate your quick response and suggestion. They are very helpful. Did you trim the adaptor and primer sequences before alignment? Did your merge forward and reverse reads? Which one goes first?

Thanks,
Wendy

**Comment by guodudou2** [comment]
Hi Matt,

Do you know how your amplicon library was prepared?  I found both 3' and 5' primers appearing in either forward or reserve reads fastq file.  Is this normal? After we cut the primer, can we just align with the trimmed R1 and R2 as they are?

Thanks,
Wendy

**Comment by eldrid01** [comment]
Hi Wendy,

Trimming adapter or primer sequences is not needed and the pipeline expects the primer sequences to be part of the reads. 

You need to specify the coordinates of both the amplicon, i.e. the region including the primer sequences, and the target, i.e. the region between the primer sequences - if you look at the example from one of our amplicon coordinate files given in the README you can see that our primer sequences are each around 21 nucleotides.

We would expect to see the 5' primer at the beginning of one of the reads and the 3' primer at the beginning of the other read from the same pair. If the read length is longer than the amplicon then you'll see both the 5' and 3' primers in each reads and the reads may run into adapter. Using an aligner such as bwa-mem or bwa-mem2 that can align part of a read is  important if that's the case.

I think the library prep for most of our amplicon sequencing follows the original TAm-seq approach described here:
https://pubmed.ncbi.nlm.nih.gov/22649089
https://doi.org/10.1126/scitranslmed.3003726

Best wishes,
Matt

**Comment by guodudou2** [comment]
Thanks Matt! @eldrid01 

I aligned my data to reference genome, now is running your pipeline on just one amplicon sample.  I encountered java memory problem for the following step. Do you know how to change the setting of `JAVA_OPTS="-Xmx3904m"`, or maybe it is due to other problem?

Thanks!

```
#!/bin/bash
  JAVA_OPTS="-Xmx3904m" pileup-counts \
      --id "NRAS_Exon3" \
      --input NRAS_Exon3.1.bam \
      --amplicon-intervals targets.1.bed \
      --reference-sequence genome.fa \
      --output "NRAS_Exon3.1.pileup_counts.txt" \
      --minimum-mapping-quality 1 \
      --minimum-base-quality 10
```


#### Error message is as below
Command exit status:
  1

Command output:
  2023-09-29 14:36:03,020 INFO  PileupCounts - org.cruk.ampliconseq.PileupCounts (ampliconseq 1.0-SNAPSHOT)
  2023-09-29 14:36:03,083 INFO  PileupCounts - Amplicon: NRAS_Exon3

Command error:
  2023-09-29 14:36:03,020 INFO  PileupCounts - org.cruk.ampliconseq.PileupCounts (ampliconseq 1.0-SNAPSHOT)
  2023-09-29 14:36:03,083 INFO  PileupCounts - Amplicon: NRAS_Exon3
  Exception in thread "main" java.lang.OutOfMemoryError: Java heap space: failed reallocation of scalar replaced objects


#### my config file is as below
```
params {
    samples               = "/data/wendy/project/braf/amplicon01/info/samples.tsv"
    bamDir                = "/data/wendy/project/braf/amplicon01/fastq"
    amplicons             = "/data/wendy/project/braf/amplicon01/info/amplicons.tsv"
    referenceGenomeFasta  = "/data/wendy/reference/igenomes2023/Homo_sapiens/NCBI/GRCh38/Sequence/WholeGenomeFasta/genome.fa"
    vepAnnotation         = true
    vepCacheDir           = "/data/wendy/vep_cache/homo_sapiens/108_GRCh38"
    vepSpecies            = "homo_sapiens"
    vepAssembly           = "GRCh38"
    outputDir             = "/data/wendy/project/braf/amplicon01/results"
    variantCaller         = "VarDict"
    minimumAlleleFraction = 0.01
}

profiles {
    myprofile {
        process.executor = 'local'
        executor {
            cpus = 8
            memory = 600.GB
        }
        docker.enabled = true
    }
}

```

#### nextflow running script is as below
`nextflow run crukci-bioinformatics/ampliconseq -c /data/wendy/project/braf/amplicon01/config/nras.config -resume -profile myprofile`

**Comment by eldrid01** [comment]
Hi Wendy,

The pileup_counts process is configured in the pipeline to run in 4GB RAM and usually that is sufficient. It is configured to retry some tasks, including pileup_counts, that fail with certain exit codes but those we've specified in the pipeline are more specifically to do with out of memory errors, e.g. 137. The exit code you get depends largely on the system you're running on and clearly in this case, you're not getting the retry behaviour.

You can override the default memory setting for the pileup_counts process and set it to a larger heap allocation by adding the following to your nextflow.config file.

process {
    withName: pileup_counts {
        memory = 8.GB
    }
}

I've not seen this error before but hopefully that will do the trick.

**Comment by guodudou2** [comment]
Thanks, Matt @eldrid01 ! Your way solved my problem. Somehow, no variant has been called. When I look at the bam file in IGV, I see the interested position is actually heterozygous. Do you think which step may cause this problem? Based on IGV, I think the amplicon and target positions may be different from what I specified as input. Do you think this is the reason?

![image](https://github.com/crukci-bioinformatics/ampliconseq/assets/137312732/2342fd5c-359e-469f-8f83-341f636ca387)
![image](https://github.com/crukci-bioinformatics/ampliconseq/assets/137312732/29767a66-118e-47c5-9b2f-005a44168bf5)

Thanks,
Wendy

**Comment by eldrid01** [comment]
Hi Wendy, I'm struggling a bit to understand what is going on with your read alignments in this screenshot.

It looks like your amplicon is approximately 270bp long and starts at around chr1:114,713,800 and ends at chr1:114,714,070. It also looks like you have 100bp paired end reads. The reads that start from the 5' primer align fairly nicely to the chr1:114,713,800 as I would expect. And you have very high depth sequencing, around 1,000,000x coverage. Is that correct?

If so, then the variant of interest appears to be in the gap between the paired end reads. 

However, clearly you have some reads which are aligning in this gap region in the middle of the amplicon, around 1100x coverage and these show the variant to be heterozygous.

Also, the pipeline really does not expect the reads to have been trimmed and it looks like your reads have been. The names of the BAM files suggest some quality trimming has been performed.

Normally I would expect to see the read pairs aligning along their whole length and aligning consistently to the same position for each amplicon. So in your case I would expect to see two big rectangular blocks of aligned reads, one where the reads align from 114,713-800 - 114,713,900 and the other for the 3' primer aligning to 114,713,970 - 114,714,070. And no reads aligning elsewhere including in the gap between these (114,713,900 - 114,713,970).

**Comment by guodudou2** [comment]
Hi Matt @eldrid01, you are right, this is something I am very confused too. I have been double checking with Yale sequencing center to discuss why we see reads with lower coverage (~1000) in the middle.  I used untrimmed (neither quality nor primer) as input to your pipeline. I aligned the reads after qualify trimming only for IGV purpose. I think the unexpected library prep procedure may cause the problem. I will update after I hear back from Yale genomic center.  In addition, I am checking whether MUTECT2 can call variants with amplicon seq and wondering why you choose hyplotypecaller over mutect2? 

Thanks,
Wendy

**Comment by eldrid01** [comment]
Hi Wendy,

Originally the pipeline was developed for a group here at the Cancer Research UK Cambridge Institute who were looking for germline mutations in various genes and also a clonal p53 mutation that was expected to have a high variant allele frequency. The HaplotypeCaller seemed like the best choice at the time but it was a while back and although we did look at Mutect2 it was marked as 'experimental' in the GATK documentation for single sample (non-somatic) variant calling back then.

When the pipeline was subsequently extended for lower allele fraction variant calling, the research group we were working with needed a caller that could make calls down to a VAF of around 0.01 and were particularly interested in VarDict, so that's what got added as an option, as well as the background noise filtering. There's no reason why Mutect2 couldn't be offered as an option and I've thought of doing so but there's not yet been a strong enough pull from the research groups here that use the pipeline.

The pipeline is designed to work with panels of amplicons that tile across genes of interest and so might be overlapping. The pipeline identifies the reads associated with each amplicon and only uses these for calling variants in the target (non-primer) region of the amplicon. That way, the primer sequence parts of the reads never get used in calling variants. For your case (in the IGV screenshot) I would not expect any of the reads covering the variant locus to be assigned to the amplicon and the pipeline would not be able to call that variant as it would see no coverage there.

I think you're right to check with the sequencing centre about the library prep. It's a very small proportion of the reads that align unexpectedly(?) in the gap between the vast majority of the paired reads which I suppose raises questions about the efficiency of the amplicon primer design for targeting that locus specifically.

Best wishes,
Matt
